src/realm/db.cpp

```diff
@@ -78,304 +78,192 @@ namespace {
     {
-        // do not call this if the buffer is full!
-        uint_fast32_t idx = get(last()).next;
-        return idx;
-    }
-
-    ReadCount& get_next() noexcept
+        auto& r = get(idx);
+        if (idx == oldest)
+            oldest = r.newer;
+        if (idx == newest)
+            newest = r.older; // can this happen?
+        if (r.older != nil)
+            get(r.older).newer = r.newer;
+        if (r.newer != nil)
+            get(r.newer).older = r.older;
+
+        // add to freelist
+        r.newer = nil;
+        r.older = freelist;
+        freelist = idx;
+    }
+
+    // This method re-initialises the last used VersionList entry to hold a new entry.
+    // Precondition: This should *only* be done if the caller has established that she
+    // is the only thread/process that has access to the VersionList. It is currently
+    // called from init_versioning(), which is called by DB::open() under the
+    // condition that it is the session initiator and under guard by the control mutex,
+    // thus ensuring the precondition.
+    // It is most likely not suited for any other use.
+    ReadCount& init_versioning() noexcept
     {
-        REALM_ASSERT(!is_full());
-        return data[next()];
+        int num_entries = entries;
+        entries = 0;
+        newest = oldest = freelist = nil;
+        expand_to(num_entries);
+        ReadCount& r = allocate_entry();
+        r.count_live = r.count_frozen = 0;
+        return r;
     }
 
-    void use_next() noexcept
+    bool is_full() const noexcept
     {
-        atomic_dec(get_next().count); // .store_release(0);
-        put_pos.store(uint32_t(next()), std::memory_order_release);
+        return freelist == nil;
     }
 
-    void cleanup() noexcept
+    void purge_versions(uint64_t& oldest_v, uint64_t& oldest_live_v, TopRefMap& top_refs,
+                        int& num_unreachable) noexcept
     {
-        // invariant: entry held by put_pos has count > 1.
-        // std::cout << "cleanup: from " << old_pos << " to " << put_pos.load_relaxed();
-        // dump();
-        while (old_pos.load(std::memory_order_relaxed) != put_pos.load(std::memory_order_relaxed)) {
-            const ReadCount& r = get(old_pos.load(std::memory_order_relaxed));
-            if (!atomic_one_if_zero(r.count))
-                break;
-            auto next_ndx = get(old_pos.load(std::memory_order_relaxed)).next;
-            old_pos.store(next_ndx, std::memory_order_relaxed);
+        // run through list of versions and reclaim unused entries.
+        // buld a map of reachable versions on the way.
+        // there is always at least one live entry
+        auto it = oldest;
+        bool looking_for_oldest_live = true;
+        bool have_seen_a_reachable_version = false;
+        REALM_ASSERT(oldest != nil);
+        while (it != nil) {
+            auto& r = get(it);
+            it = r.newer;
+            if (looking_for_oldest_live && r.count_live != 0) {
```

Comment:
It is likely faster to do 2 linear scans, one to find the lowest live and the lowest reachable versions, then another pass to act on those numbers. If you do this, I think you can completely eliminate the whole doubly linked list, and shrink the size of ReadCount by 8 bytes, making it a nice power of 2 (you can embed the freelist link in a field that is unused in free links). Or you can just get rid of the freelist and just grab the lowest available entry in `allocate_entry` since commit is already linear in the size of this list.